Fanspeed deprecated warnings on Dev

### The problem

Have noticed warnings of deprecated fanspeed even though I haven't programed a fan speed.
Only using a binary fan.

Also have noticed for a while now incompatable  pointer types as seen in log below.

### Which version of ESPHome has the issue?

V2021.9.0-dev

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

2021.9.3

### What platform are you using?

ESP32

### Board

multiple

### Component causing the issue

fan

### Example YAML snippet

```yaml
esphome:
  name: test
  platform: ESP32
  board: esp-wrover-kit
  
output:
  - id: relay_1
    platform: gpio
    pin: GPIO5
    inverted: true

fan:
   - platform: binary
     output: relay_1
     name: "test"
```


### Anything in the logs that might be useful for us?

```txt
INFO Reading configuration /config/esphome/test.yaml...
INFO Generating C++ source...
INFO Core config or version changed, cleaning build files...
INFO Deleting /data/test/.pioenvs
INFO Compiling app...
INFO Running:  platformio run -d /config/esphome/test
Processing test (board: esp-wrover-kit; framework: arduino; platform: platformio/espressif32@3.2.0)
--------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
PACKAGES: 
 - framework-arduinoespressif32 3.10006.210326 (1.0.6) 
 - tool-esptoolpy 1.30000.201119 (3.0.0) 
 - toolchain-xtensa32 2.50200.97 (5.2.0)
Dependency Graph
|-- <ESPmDNS> 1.0
|   |-- <WiFi> 1.0
Compiling /data/test/.pioenvs/test/src/esphome/components/binary/fan/binary_fan.cpp.o
Compiling /data/test/.pioenvs/test/src/esphome/components/fan/automation.cpp.o
Compiling /data/test/.pioenvs/test/src/esphome/components/fan/fan_helpers.cpp.o
Compiling /data/test/.pioenvs/test/src/esphome/components/fan/fan_state.cpp.o
Compiling /data/test/.pioenvs/test/src/esphome/components/gpio/output/gpio_binary_output.cpp.o
Compiling /data/test/.pioenvs/test/src/esphome/components/output/automation.cpp.o
Compiling /data/test/.pioenvs/test/src/esphome/components/output/float_output.cpp.o
Compiling /data/test/.pioenvs/test/src/esphome/core/application.cpp.o
In file included from src/esphome/components/fan/fan_helpers.cpp:2:0:
src/esphome/components/fan/fan_helpers.h:7:73: warning: 'FanSpeed' is deprecated: FanSpeed is deprecated. [-Wdeprecated-declarations]
 FanSpeed speed_level_to_enum(int speed_level, int supported_speed_levels);
                                                                         ^
In file included from src/esphome/components/fan/fan_helpers.h:2:0,
                 from src/esphome/components/fan/fan_helpers.cpp:2:
src/esphome/components/fan/fan_state.h:13:57: note: declared here
 enum ESPDEPRECATED("FanSpeed is deprecated.", "2021.9") FanSpeed {
                                                         ^
In file included from src/esphome/components/fan/fan_helpers.cpp:2:0:
src/esphome/components/fan/fan_helpers.h:8:67: warning: 'FanSpeed' is deprecated: FanSpeed is deprecated. [-Wdeprecated-declarations]
 int speed_enum_to_level(FanSpeed speed, int supported_speed_levels);
                                                                   ^
In file included from src/esphome/components/fan/fan_helpers.h:2:0,
                 from src/esphome/components/fan/fan_helpers.cpp:2:
src/esphome/components/fan/fan_state.h:13:57: note: declared here
 enum ESPDEPRECATED("FanSpeed is deprecated.", "2021.9") FanSpeed {
                                                         ^
src/esphome/components/fan/fan_helpers.cpp:8:73: warning: 'FanSpeed' is deprecated: FanSpeed is deprecated. [-Wdeprecated-declarations]
 FanSpeed speed_level_to_enum(int speed_level, int supported_speed_levels) {
                                                                         ^
In file included from src/esphome/components/fan/fan_helpers.h:2:0,
                 from src/esphome/components/fan/fan_helpers.cpp:2:
src/esphome/components/fan/fan_state.h:13:57: note: declared here
 enum ESPDEPRECATED("FanSpeed is deprecated.", "2021.9") FanSpeed {
                                                         ^
src/esphome/components/fan/fan_helpers.cpp: In function 'esphome::fan::FanSpeed esphome::fan::speed_level_to_enum(int, int)':
src/esphome/components/fan/fan_helpers.cpp:11:22: warning: 'FanSpeed' is deprecated: FanSpeed is deprecated. [-Wdeprecated-declarations]
   return static_cast<FanSpeed>(legacy_level - 1);  // NOLINT(clang-diagnostic-deprecated-declarations)
                      ^
In file included from src/esphome/components/fan/fan_helpers.h:2:0,
                 from src/esphome/components/fan/fan_helpers.cpp:2:
src/esphome/components/fan/fan_state.h:13:57: note: declared here
 enum ESPDEPRECATED("FanSpeed is deprecated.", "2021.9") FanSpeed {
                                                         ^
src/esphome/components/fan/fan_helpers.cpp: At global scope:
src/esphome/components/fan/fan_helpers.cpp:15:67: warning: 'FanSpeed' is deprecated: FanSpeed is deprecated. [-Wdeprecated-declarations]
 int speed_enum_to_level(FanSpeed speed, int supported_speed_levels) {
                                                                   ^
In file included from src/esphome/components/fan/fan_helpers.h:2:0,
                 from src/esphome/components/fan/fan_helpers.cpp:2:
src/esphome/components/fan/fan_state.h:13:57: note: declared here
 enum ESPDEPRECATED("FanSpeed is deprecated.", "2021.9") FanSpeed {
                                                         ^
In file included from src/esphome/components/fan/fan_state.cpp:2:0:
src/esphome/components/fan/fan_helpers.h:7:73: warning: 'FanSpeed' is deprecated: FanSpeed is deprecated. [-Wdeprecated-declarations]
 FanSpeed speed_level_to_enum(int speed_level, int supported_speed_levels);
                                                                         ^
In file included from src/esphome/components/fan/fan_state.cpp:1:0:
src/esphome/components/fan/fan_state.h:13:57: note: declared here
 enum ESPDEPRECATED("FanSpeed is deprecated.", "2021.9") FanSpeed {
                                                         ^
In file included from src/esphome/components/fan/fan_state.cpp:2:0:
src/esphome/components/fan/fan_helpers.h:8:67: warning: 'FanSpeed' is deprecated: FanSpeed is deprecated. [-Wdeprecated-declarations]
 int speed_enum_to_level(FanSpeed speed, int supported_speed_levels);
                                                                   ^
In file included from src/esphome/components/fan/fan_state.cpp:1:0:
src/esphome/components/fan/fan_state.h:13:57: note: declared here
 enum ESPDEPRECATED("FanSpeed is deprecated.", "2021.9") FanSpeed {
                                                         ^
Compiling /data/test/.pioenvs/test/src/esphome/core/application_esp32.cpp.o
Compiling /data/test/.pioenvs/test/src/esphome/core/application_esp8266.cpp.o
Compiling /data/test/.pioenvs/test/src/esphome/core/color.cpp.o
Compiling /data/test/.pioenvs/test/src/esphome/core/component.cpp.o
Compiling /data/test/.pioenvs/test/src/esphome/core/controller.cpp.o
Compiling /data/test/.pioenvs/test/src/esphome/core/esphal.cpp.o
Compiling /data/test/.pioenvs/test/src/esphome/core/helpers.cpp.o
Compiling /data/test/.pioenvs/test/src/esphome/core/log.cpp.o
Compiling /data/test/.pioenvs/test/src/esphome/core/preferences.cpp.o
Compiling /data/test/.pioenvs/test/src/esphome/core/scheduler.cpp.o
Compiling /data/test/.pioenvs/test/src/esphome/core/util.cpp.o
Compiling /data/test/.pioenvs/test/src/main.cpp.o
Generating partitions /data/test/.pioenvs/test/partitions.bin
Compiling /data/test/.pioenvs/test/lib789/WiFi/ETH.cpp.o
Compiling /data/test/.pioenvs/test/lib789/WiFi/WiFi.cpp.o
Compiling /data/test/.pioenvs/test/lib789/WiFi/WiFiAP.cpp.o
Compiling /data/test/.pioenvs/test/lib789/WiFi/WiFiClient.cpp.o
In file included from src/esphome.h:4:0,
                 from src/main.cpp:3:
src/esphome/components/fan/fan_helpers.h:7:73: warning: 'FanSpeed' is deprecated: FanSpeed is deprecated. [-Wdeprecated-declarations]
 FanSpeed speed_level_to_enum(int speed_level, int supported_speed_levels);
                                                                         ^
In file included from src/esphome/components/binary/fan/binary_fan.h:5:0,
                 from src/esphome.h:2,
                 from src/main.cpp:3:
src/esphome/components/fan/fan_state.h:13:57: note: declared here
 enum ESPDEPRECATED("FanSpeed is deprecated.", "2021.9") FanSpeed {
                                                         ^
In file included from src/esphome.h:4:0,
                 from src/main.cpp:3:
src/esphome/components/fan/fan_helpers.h:8:67: warning: 'FanSpeed' is deprecated: FanSpeed is deprecated. [-Wdeprecated-declarations]
 int speed_enum_to_level(FanSpeed speed, int supported_speed_levels);
                                                                   ^
In file included from src/esphome/components/binary/fan/binary_fan.h:5:0,
                 from src/esphome.h:2,
                 from src/main.cpp:3:
src/esphome/components/fan/fan_state.h:13:57: note: declared here
 enum ESPDEPRECATED("FanSpeed is deprecated.", "2021.9") FanSpeed {
                                                         ^
Compiling /data/test/.pioenvs/test/lib789/WiFi/WiFiGeneric.cpp.o
Compiling /data/test/.pioenvs/test/lib789/WiFi/WiFiMulti.cpp.o
Compiling /data/test/.pioenvs/test/lib789/WiFi/WiFiSTA.cpp.o
Compiling /data/test/.pioenvs/test/lib789/WiFi/WiFiScan.cpp.o
Compiling /data/test/.pioenvs/test/lib789/WiFi/WiFiServer.cpp.o
Compiling /data/test/.pioenvs/test/lib789/WiFi/WiFiUdp.cpp.o
Compiling /data/test/.pioenvs/test/libe94/ESPmDNS/ESPmDNS.cpp.o
Archiving /data/test/.pioenvs/test/libFrameworkArduinoVariant.a
Compiling /data/test/.pioenvs/test/FrameworkArduino/Esp.cpp.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/FunctionalInterrupt.cpp.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/HardwareSerial.cpp.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/IPAddress.cpp.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/IPv6Address.cpp.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/MD5Builder.cpp.o
Archiving /data/test/.pioenvs/test/lib789/libWiFi.a
Compiling /data/test/.pioenvs/test/FrameworkArduino/Print.cpp.o
Archiving /data/test/.pioenvs/test/libe94/libESPmDNS.a
Compiling /data/test/.pioenvs/test/FrameworkArduino/Stream.cpp.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/StreamString.cpp.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/WMath.cpp.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/WString.cpp.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/base64.cpp.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/cbuf.cpp.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/esp32-hal-adc.c.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/esp32-hal-bt.c.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/esp32-hal-cpu.c.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/esp32-hal-dac.c.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/esp32-hal-gpio.c.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/esp32-hal-i2c.c.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/esp32-hal-ledc.c.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/esp32-hal-log.c.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/esp32-hal-matrix.c.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/esp32-hal-misc.c.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/esp32-hal-psram.c.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/esp32-hal-rmt.c.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/esp32-hal-sigmadelta.c.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/esp32-hal-spi.c.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/esp32-hal-time.c.o
/root/.platformio/packages/framework-arduinoespressif32/cores/esp32/esp32-hal-spi.c: In function 'spiTransferBytesNL':
/root/.platformio/packages/framework-arduinoespressif32/cores/esp32/esp32-hal-spi.c:922:39: warning: initialization from incompatible pointer type [-Wincompatible-pointer-types]
                 uint8_t * last_out8 = &result[c_longs-1];
                                       ^
/root/.platformio/packages/framework-arduinoespressif32/cores/esp32/esp32-hal-spi.c:923:40: warning: initialization from incompatible pointer type [-Wincompatible-pointer-types]
                 uint8_t * last_data8 = &last_data;
                                        ^
Compiling /data/test/.pioenvs/test/FrameworkArduino/esp32-hal-timer.c.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/esp32-hal-touch.c.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/esp32-hal-uart.c.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/libb64/cdecode.c.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/libb64/cencode.c.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/main.cpp.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/stdlib_noniso.c.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/wiring_pulse.c.o
Compiling /data/test/.pioenvs/test/FrameworkArduino/wiring_shift.c.o
Archiving /data/test/.pioenvs/test/libFrameworkArduino.a
Linking /data/test/.pioenvs/test/firmware.elf
Retrieving maximum program size /data/test/.pioenvs/test/firmware.elf
Checking size /data/test/.pioenvs/test/firmware.elf
RAM:   [          ]   4.3% (used 14128 bytes from 327680 bytes)
Flash: [==        ]  15.5% (used 284842 bytes from 1835008 bytes)
Building /data/test/.pioenvs/test/firmware.bin
========================= [SUCCESS] Took 8.60 seconds =========================
INFO Successfully compiled program.
```


### Additional information

_No response_

--- a/esphome/components/api/api_connection.cpp
+++ b/esphome/components/api/api_connection.cpp
@@ -214,6 +214,9 @@ void APIConnection::cover_command(const CoverCommandRequest &msg) {
 #endif
 
 #ifdef USE_FAN
+// Shut-up about usage of deprecated speed_level_to_enum/speed_enum_to_level functions for a bit.
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
 bool APIConnection::send_fan_state(fan::FanState *fan) {
   if (!this->state_subscription_)
     return false;
@@ -262,13 +265,13 @@ void APIConnection::fan_command(const FanCommandRequest &msg) {
     // Prefer level
     call.set_speed(msg.speed_level);
   } else if (msg.has_speed) {
-    // NOLINTNEXTLINE(clang-diagnostic-deprecated-declarations)
     call.set_speed(fan::speed_enum_to_level(static_cast<fan::FanSpeed>(msg.speed), traits.supported_speed_count()));
   }
   if (msg.has_direction)
     call.set_direction(static_cast<fan::FanDirection>(msg.direction));
   call.perform();
 }
+#pragma GCC diagnostic pop
 #endif
 
 #ifdef USE_LIGHT

--- a/esphome/components/fan/fan_helpers.cpp
+++ b/esphome/components/fan/fan_helpers.cpp
@@ -4,14 +4,15 @@
 namespace esphome {
 namespace fan {
 
-// NOLINTNEXTLINE(clang-diagnostic-deprecated-declarations)
+// This whole file is deprecated, don't warn about usage of deprecated types in here.
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+
 FanSpeed speed_level_to_enum(int speed_level, int supported_speed_levels) {
   const auto speed_ratio = static_cast<float>(speed_level) / (supported_speed_levels + 1);
   const auto legacy_level = clamp<int>(static_cast<int>(ceilf(speed_ratio * 3)), 1, 3);
-  return static_cast<FanSpeed>(legacy_level - 1);  // NOLINT(clang-diagnostic-deprecated-declarations)
+  return static_cast<FanSpeed>(legacy_level - 1);
 }
 
-// NOLINTNEXTLINE(clang-diagnostic-deprecated-declarations)
 int speed_enum_to_level(FanSpeed speed, int supported_speed_levels) {
   const auto enum_level = static_cast<int>(speed) + 1;
   const auto speed_level = roundf(enum_level / 3.0f * supported_speed_levels);

--- a/esphome/components/fan/fan_helpers.h
+++ b/esphome/components/fan/fan_helpers.h
@@ -4,8 +4,16 @@
 namespace esphome {
 namespace fan {
 
+// Shut-up about usage of deprecated FanSpeed for a bit.
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+
+ESPDEPRECATED("FanSpeed and speed_level_to_enum() are deprecated.", "2021.9")
 FanSpeed speed_level_to_enum(int speed_level, int supported_speed_levels);
+ESPDEPRECATED("FanSpeed and speed_enum_to_level() are deprecated.", "2021.9")
 int speed_enum_to_level(FanSpeed speed, int supported_speed_levels);
+
+#pragma GCC diagnostic pop
 
 }  // namespace fan
 }  // namespace esphome

--- a/esphome/components/fan/fan_state.cpp
+++ b/esphome/components/fan/fan_state.cpp
@@ -67,6 +67,8 @@ void FanStateCall::perform() const {
   this->state_->state_callback_.call();
 }
 
+// This whole method is deprecated, don't warn about usage of deprecated methods inside of it.
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
 FanStateCall &FanStateCall::set_speed(const char *legacy_speed) {
   const auto supported_speed_count = this->state_->get_traits().supported_speed_count();
   if (strcasecmp(legacy_speed, "low") == 0) {

--- a/esphome/components/mqtt/mqtt_fan.cpp
+++ b/esphome/components/mqtt/mqtt_fan.cpp
@@ -100,6 +100,7 @@ bool MQTTFanComponent::publish_state() {
   auto traits = this->state_->get_traits();
   if (traits.supports_speed()) {
     const char *payload;
+    // NOLINTNEXTLINE(clang-diagnostic-deprecated-declarations)
     switch (fan::speed_level_to_enum(this->state_->speed, traits.supported_speed_count())) {
       case FAN_SPEED_LOW: {  // NOLINT(clang-diagnostic-deprecated-declarations)
         payload = "low";

--- a/esphome/components/web_server/web_server.cpp
+++ b/esphome/components/web_server/web_server.cpp
@@ -397,6 +397,7 @@ std::string WebServer::fan_json(fan::FanState *obj) {
     const auto traits = obj->get_traits();
     if (traits.supports_speed()) {
       root["speed_level"] = obj->speed;
+      // NOLINTNEXTLINE(clang-diagnostic-deprecated-declarations)
       switch (fan::speed_level_to_enum(obj->speed, traits.supported_speed_count())) {
         case fan::FAN_SPEED_LOW:  // NOLINT(clang-diagnostic-deprecated-declarations)
           root["speed"] = "low";

